java/arcs/core/host/ArcState.kt

```diff
@@ -10,97 +10,201 @@
+            "${state.name}|$_cause"
+        } else {
+            state.name
+        }
+    }
 
-    /**
-     * The [Arc] could not be started for some reason, usually due to the failure of one of its
-     * [Particle]s to be instantiated properly.
-     */
-    Error,
+    enum class State {
+        /**
+         * The [Arc] is in between [Stopped] and [Running] states, with some [ArcHost]s or
+         * [Particle]s started, and others not.
+         */
+        Indeterminate,
 
-    /**
-     * [Deleted] implies [Stopped], but furthermore, subsequent attempts to restart this
-     * [Arc] will fail, and potentially any data used by the [Arc] may be reclaimed.
-     */
-    Deleted
+        /**
+         * The [Plan.Partition] is running on this particular [ArcHost]. An [Arc] is considered
+         * running if even one of its [ArcHost]s are in the [Running] state.
+         */
+        Running,
+
+        /**
+         * The [Plan.Partition] running on this particular [ArcHost] has been stopped, either
+         * because it was explicitly stopped, or the [ArcHost] was restarted, but the arc has
+         * not been resurrected yet. An [Arc] is considered stopped if and only if all of its
+         * participating [ArcHost]s have reached the [Stopped] state.
+         */
+        Stopped,
+
+        /**
+         * This particular [ArcHost] has never started an [Arc] for this [Plan.Partition] yet.
+         */
+        NeverStarted,
+
+        /**
+         * The [Arc] could not be started for some reason, usually due to the failure of one of its
+         * [Particle]s to be instantiated properly.
+         */
+        Error,
+
+        /**
+         * [Deleted] implies [Stopped], but furthermore, subsequent attempts to restart this
+         * [Arc] will fail, and potentially any data used by the [Arc] may be reclaimed.
+         */
+        Deleted
+    }
+
+    companion object {
+        val Indeterminate = ArcState(State.Indeterminate)
+        val Running = ArcState(State.Running)
+        val Stopped = ArcState(State.Stopped)
+        val NeverStarted = ArcState(State.NeverStarted)
+        val Error = ArcState(State.Error)
+        val Deleted = ArcState(State.Deleted)
+
+        /**
+         * Creates an ArcState.Error instance with an exception attached. Note that the exception
+         * is not included in equality comparisons, so two Error instances with different exceptions
+         * will still be considered equal, both to each other and to the singleton Error defined
+         * above.
+         */
+        fun errorWith(cause: Exception? = null) = ArcState(State.Error).apply { _cause = cause }
+
+        /**
+         * Creates an ArcState instance from the given string. For the Error case, this may have
+         * a reconstructed exception containing the original exception's toString result.
+         */
+        fun fromString(value: String): ArcState {
+            val parts = value.split("|", limit = 2)
```

Comment:
optional: An inline class is a good way to make such 'String' args more typesafe, see ArcStateChangeRegistration.kt.  I mention this because you may want to add more and more details to this serialization in the future, like nested cause support, and support for stack trace encoding, e.g.

```
inline class SerializedState(internal val serializeState: String) {
  constructor(state: State) = this(state.name)
  constructor(exception: SerializedException) = this("${Error.name}|$exception")
  val state: State get() = State.valueOf(serializeState.split("|", limit = 2)[0])
  val exception: DeserializedException get() = SerializedException(serializeState.split("|", limit = 2)[1]).exception
}

inline class SerializedException(internal val serializedException: String) {
   constructor(exceptionName: String, message: String, cause: SerializedException, stackTrace: StackTrace) = this(serialize(exceptionName, message, cause, stackTrace))

  fun serialize(name: String, message: String, cause: SerializedException, stackTrace: StackTrace) = "$name~$message~$stackTrace~$cause"
 
  val exception: DeserializedException get() {
    val parts = serializedException.split("~", limit = 4)
      DeserializedException(parts[0], parts[1], parts[2], parts[3])
  }
}
```

The general idea is to keep them as Strings, by put them behind a type-safe helper that you can treat like class instances, and add serialization/deserialization logic to.

Reply:
Excellent idea, will do in a follow-up CL.